relay-profiling/src/android.rs

```diff
@@ -99,29 +122,109 @@ impl AndroidProfile {
+        let start_time = new_profile.profile.start_time;
+
+        new_profile.profile.events.retain_mut(|event| {
+            let event_timestamp = get_timestamp(clock, start_time, event.time);
+            if transaction.relative_start_ns <= event_timestamp
+                && event_timestamp <= transaction.relative_end_ns
+            {
+                let relative_start = match clock {
+                    Clock::Cpu => Duration::from_millis(transaction.relative_cpu_start_ms),
+                    Clock::Wall | Clock::Dual | Clock::Global => {
+                        Duration::from_nanos(transaction.relative_start_ns)
+                    }
+                };
+                event.time = substract_to_timestamp(event.time, relative_start);
+                true
+            } else {
+                false
+            }
+        });
+
+        if new_profile.profile.events.is_empty() {
+            continue;
+        }
+
+        match serde_json::to_vec(&new_profile) {
+            Ok(payload) => items.push(payload),
+            Err(_) => {
+                return Err(ProfileError::CannotSerializePayload);
+            }
+        };
     }
 
-    match serde_json::to_vec(&profile) {
-        Ok(payload) => Ok(payload),
-        Err(_) => Err(ProfileError::CannotSerializePayload),
+    Ok(items)
+}
+
+fn substract_to_timestamp(time: Time, duration: Duration) -> Time {
+    match time {
+        Time::Global(time) => Time::Global(time - duration),
+        Time::Monotonic {
+            cpu: Some(time),
+            wall: None,
+        } => Time::Monotonic {
+            cpu: time
+                .checked_sub(duration)
+                .or_else(|| Some(Duration::default())),
+            wall: None,
+        },
+        Time::Monotonic {
+            wall: Some(time),
+            cpu: None,
+        } => Time::Monotonic {
+            wall: time
+                .checked_sub(duration)
+                .or_else(|| Some(Duration::default())),
+            cpu: None,
+        },
+        Time::Monotonic {
+            cpu: Some(cpu),
+            wall: Some(wall),
+        } => Time::Monotonic {
+            cpu: cpu
+                .checked_sub(duration)
+                .or_else(|| Some(Duration::default())),
+            wall: wall
+                .checked_sub(duration)
+                .or_else(|| Some(Duration::default())),
+        },
+        _ => unimplemented!(),
     }
 }
 
 fn get_timestamp(clock: Clock, start_time: DateTime<Utc>, event_time: Time) -> u64 {
     match (clock, event_time) {
         (Clock::Global, Time::Global(time)) => {
-            time.as_nanos() as u64 - start_time.timestamp_nanos() as u64
+            let time_ns = time.as_nanos() as u64;
+            let start_time_ns = start_time.timestamp_nanos() as u64;
+            if time_ns >= start_time_ns {
+                time.as_nanos() as u64 - start_time.timestamp_nanos() as u64
+            } else {
+                0
+            }
```

Comment:
`time_ns` and `start_time_ns` are already calculated just above, no need to redo it:
```suggestion
            if time_ns >= start_time_ns {
                time_ns - start_time_ns
            } else {
                0
            }
```

Reply:
Indeed.